### ISSUE TITLE
API Proposal around Roslyn Analyser and F# compiler RFC RequireArgumentNameAttribute

### ISSUE THREAD

**Issue by smoothdeveloper** [dotnet/runtime#51451]
## Background and Motivation

An approved F# suggestion is being drafted [as an RFC](https://github.com/fsharp/fslang-design/blob/master/drafts/FS-1095-requirenamedargumentattribute.md) that enforces method call sites to carry argument name when a method is defined with `[RequireArgumentName]`.

And a similar request was made for C#, but rejected on the ground of "analyser fits the bill".

I'd like the BCL & runtime team to consider if it makes sense to bake the attribute into the BCL rather than going with it in FSharp.Core and letting the non F# users deal with more artisanal approach (anyone coming with their own analyser and attribute couple) while the motives applying to F# in terms of API design apply equally well into the whole dotnet ecosystem (in my opinion).

* suggestion for C#: https://github.com/dotnet/csharplang/discussions/1005 rejected for the analyser route
* RFC for F#: https://github.com/fsharp/fslang-design/blob/master/drafts/FS-1095-requirenamedargumentattribute.md
* discussions so far on the F# RFC https://github.com/fsharp/fslang-design/discussions/538

If there is interest for the attribute to be part of BCL (to cut the concern of what I called artisanal approach above), I'll bring more content to this API proposal based on work so far on the F# RFC and the expected formalism for API proposals.

cc'ing two suspects that you may engage with at Microsoft with a bit of background: @CyrusNajmabadi & @cartermp in case they'd like to engage / nudge in a particular direction.

**Comment by smoothdeveloper** [comment]
@jkotas / @jeffschwMSFT / @CyrusNajmabadi /@cartermp any feedback on this?

Is it something we can get better scoped, pushed for later, binned within next couple of weeks?

Is the BCL fine with the "artisanal approach" of attributes/analysers couples sprouting for this or would rather have a mere canonical attribute defined in the BCL itself.

Thanks!

**Comment by jkotas** [comment]
Analyzers go through the same review process as APIs: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md. It would help if you reformat the proposal in this issue to follow the template. In particular, the proposal should include guidance on where the library authors (including the dotnet/runtime libraries) should use this attribute.

**Comment by smoothdeveloper** [comment]
@jkotas, thanks a lot for the feedback, it makes sense and I'm happy to hear that there is formal process in place for API suggestions review (it is more foggy, but still effective in F# land).

I've updated the main description, please give me feedback if the formalism can be improved before it lands in a design meeting.

Also, if those design meetings are sometimes open to external contributors for the aim of discussing the issue in detail, I'd be happy to join such meeting.

**Comment by jkotas** [comment]
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/2220 is a very similar proposal to this one. It should be included in alternative designs. Note that it does not require any attributes, it is only driven by heuristics.

**Comment by jkotas** [comment]
More existing art:

https://www.jetbrains.com/help/rider/Argument_Style.html
https://www.jetbrains.com/help/rider/Inline_Parameter_Name_Hints.html#parameter-name-hints

**Comment by smoothdeveloper** [comment]
Thanks @jkotas, added to the introductory list of references!

I think this highlights it is common concern that some people may want this enforced at callsites of API they design.

**Comment by FilipToth** [comment]
I like the first approach, Although I do feel like when you need to force named arguments in a method, that’d just bad API design. I guess might be useful in some cases like constructors for classes that need to connect to a server and require credentials etc... I would like to see the exception that gets thrown when the user doesn’t use named arguments. About the alternative design, it just doesn't belong to C#.

**Comment by smoothdeveloper** [comment]
@FilipToth 

> I would like to see the exception that gets thrown when the user doesn’t use named arguments

In F# proposed implementation, it currently looks like this, could be either warning or error depending how it feels to the BCL and compiler teams:

> The method '%s' has the 'RequireArgumentNameAttribute' attribute specified, use the named arguments syntax (e.g. 'MethodName(x = value)').

**Comment by FilipToth** [comment]
> In F# proposed implementation, it currently looks like this, could be either warning or error depending on how it feels to the BCL and compiler teams:

Sounds very reasonable. I don't think an exception needs to be thrown as this is nothing “critical” just code styling. Thanks for the clarification.

The only other minor risk I can think of is how the end-user will be notified that the method requires named arguments but that's more IntelliSense than runtime design.

As with any API additions, there will be things to tweak but this sounds great so far!

**Comment by smoothdeveloper** [comment]
Advantage of defining the attribute in the BCL in context of compiler or analyser implementation:

https://github.com/fsharp/fslang-suggestions/issues/414#issuecomment-730700801

For usage with Entity Framework:

https://github.com/fsharp/fslang-suggestions/issues/414#issuecomment-873813200

@jkotas, is this suggestion scoped for review / discussion or are there concerns so far?

**Comment by jkotas** [comment]
It is up to @dotnet/area-system-runtime to drive the decision on whether this style analyzer and accompanying attribute is worth having in the core platform.

To me, it feels very niche, it would be typically just compensating for bad API design as @FilipToth said, so it is not a clear cut.

**Comment by tannergooding** [comment]
This issue does come up from time to time but I'm not convinced that requiring the argument name to be specified is the correct approach. A related place this shows up is in various `Copy` APIs where we haven't always been consistent on whether `dest` or `src` comes first (or where users have their own notions).

I wonder how much of this could be covered without an attribute, simply by doing analysis such as:
* if you pass in a parameter named `foo` to a parameter with a different name, but one of the other parameters is named `foo` instead

**Comment by FilipToth** [comment]
Yes, this API is mostly compensating for bad design, but I think it would be beneficial. Maybe instead of throwing a warning, how about something a little more subtle? Something just enough to say: "Hey, consumers of this API. We recommend using named arguments on the call sites of this method.", but not cause any "warning-pollution" when compiling. Also, one thing to consider is that F# focuses a lot more on the functional side of programming, so in F# this API automatically introduces more value. (Please correct me if I'm wrong) Maybe just show it in Intellisense? Although, that would be a lot harder to implement. I'd love to see the more people @dotnet/area-system-runtime share their opinions, just like @jkotas already said. Then we can hopefully run this through the API review team and see how they feel about this.

**Comment by smoothdeveloper** [comment]
@tannergooding, I think having an attribute under `System.Runtime.CompilerServices` is not a huge commitment, and will enable the analyzer that care about argument name on certain methods based on heuristics to also look for that attribute.

What is needed to get this suggestion implemented in BCL?

**Comment by tannergooding** [comment]
What's the status of the F# feature. We can certainly take this to API review if and when the F# feature is looking to be finalized.

It looks like an experimental branch existed, but it was closed without being merged.

**Comment by smoothdeveloper** [comment]
Wondering if we keep dotnet going "up to 11" before we consider this, or try for 10, or keep it artisanal forever?

@tannergooding / @jaredpar / @jkotas / @terrajobst, the status is, we, in the F# community, appreciate we engage with BCL team upfront, rather than having two solutions in the long run, which pollutes and divide dotnet (albeit C# team is very stringent on supporting 0 "came first in F#" thing, a bit like sectarism, and ask F# to emit more things, to "assimilate" to later design, rather than crossbreed/support "prior F# art", which would be better for dotnet, IMO).

We've had option versus NRT (when it landed), async / task, records that can't be consumed the same, soon, it will be DU, and many subtleties where F# had to handroll, and then had to adjust when BCL and C# team moved towards progress, with no consideration for F# codebases, and comparatively less resources for language development.

So the status of F# feature won't move (unless someone else than me pushes it), until I'll be able, for win-win situation for dotnet, to make progress precisely here :)

Do you want this in the BCL, or do you want people to handroll zillion of analysers, zillions of diverging blog posts, nuances in the attribute support, no widespread consensus, no adoption, etc.

We could have dotnet support, to enrich API design, cross languages, pushed by single individual outside of microsoft, cross languages, good for long time dotnet supporters like new adopters.

What is your feedback on this current proposal? Should I make a branch & PR of runtime repo with suggested changes to move things forward in the way I intended, or that is amenable to BCL team?

The F# experimental PR was a PoC (which was quick, easy to hack the compiler), and closed for maintenance purpose, I won't make another one until we settle a discussion and next steps here.

Thanks for keeping up with me 🙂.

**Comment by smoothdeveloper** [comment]
@FilipToth 

> this API is mostly compensating for bad design

I disagree (smoothly), it is not bad design to have interspersed argument names (and even overload support for different arg names).

I feel you have some bias about language design, anchoring to languages you know.

I could come with my own biases, say "overloads are frequently bad design", or "virtual members tend to make some bad designs in naïve OO design in regular codebases", but I (pretend I) don't :)

> Also, one thing to consider is that F# focuses a lot more on the functional side of programming, so in F# this API automatically introduces more value.

Not sure, I didn't consider it like this, I actually think this feature is especially useful on heavily overloaded methods where call site being non ambiguous, through argument name, is a plus for readability, affecting equally C#, VB.NET, F#, etc.

It can be domain code, or framework code (not necessarily BCL, as I know it is such a shift to consider it could be BCL design guideline thing, beyond just a tool for opinionated API or extra safety/readability/explicitness in domain code).

Hope it clarifies the motivation a bit.

> Maybe just show it in Intellisense?

I trust BCL and C# team to define defaults for the warning being enabled or not, tooling only or not, through analyzer or in compiler (would be great to have IL analyzer API support in runtime BTW), to basically comply with their judgement, as I don't have the background to make such decisions, impacting VS and other tooling.

Adding some cc:

@vzarytovskii @dsyme @KathleenDollard @richlander

**Comment by tannergooding** [comment]
This remains same as when asked before. There's no chance for this to go forward until there's commitment from the F# team (@T-Gro) for it to ship and that it will be in a given version of the product and then we can coordinate on getting the types added if that's deemed beneficial; anything else is purely speculative and risks shipping something that isn't actually correct.

**Comment by T-Gro** [comment]
@tannergooding , @smoothdeveloper : 

This is an approved suggestion for the F# compiler to follow.
We can finalize the support once this is added to the runtime.

(I would also do the F# consumption in a polyfill-allowing way for older runtimes, only via namespace+name, not insisting on assembly identity)

**Comment by tannergooding** [comment]
@T-Gro even if its `approved in principle`, it still requires the commitment from the F# team to shipping in a given release; same as we require for C# features.

I'd also note that there should be nothing blocking F# from taking action here. That is, the polyfill approach you mention (assuming its remotely similar to how Roslyn does things) allows the feature to work even if the libraries does nothing. It is the "normal" case for most such attributes and so the expected process is:
1. The language implements the feature and it works end to end
2. The language team opens an API proposal for the attribute to avoid having to self polyfill on modern .NET
3. The API review team approves the proposal
4. The libraries team exposes the attribute in question

At that point, the feature still works end to end and you just avoid the per assembly internal duplication on modern .NET. It allows everything to light up as expected, even downlevel for devs who choose to do that; avoids ordering issues or layering complexities; and it additionally ensures that everything is shipped and tested in a way that avoids there being mistakes here (all while aligning within the same release).

I am more than happy to take this through API review, but it does need to follow the normal process. So as soon as F# is ready to commit to the feature in a given release (and ideally gets it implemented with the auto polyfill -- noting this is not required if F# truly wants to restrict it to only modern .NET or similar) then we can get up a `ready-to-review` proposal covering the exact shape and working examples that F# expects.

**Comment by T-Gro** [comment]
@tannergooding :

That makes sense.
F# team commits to ship as soon as possible, it is implemented already https://github.com/dotnet/fsharp/pull/20340

**Comment by tannergooding** [comment]
Thanks, left a couple questions on that PR that I think would be good to get answered.

I think the biggest "blocker" is determining if it needs to support the case where you only want a subset of parameters to require naming (this impacts long term shape for the general ecosystem, even if F# opts to require it for all).

I also left a few interesting scenarios described that can impact the overall UX, some of which may be F# applicable and others of which may be per language applicable.
